File: sdk/approute-sdk/javascript/tests/resources/accounts.test.ts
import { describe, it, expect } from "vitest";
import { AccountsResource } from "../../src/resources/accounts.js";
import { MockTransport } from "../support/index.js";
import {
  makeAccountListResponse,
  makeAccountTransactionPage,
} from "../factories/index.js";
import type {
  AccountListResponse,
  AccountTransactionPage,
} from "../../src/models/index.js";

describe("AccountsResource", () => {
  describe("balances()", () => {
    it("should return account balances on success", async () => {
      const fixture = makeAccountListResponse();
      const mock = new MockTransport(fixture);
      const resource = new AccountsResource(mock.transport);

      const result: AccountListResponse = await resource.balances();

      expect(result.items).toHaveLength(1);
      expect(result.items[0].currency).toBe("USD");
      expect(result.items[0].balance).toBe(1250.5);
      expect(result.items[0].available).toBe(1200.0);
      expect(result.items[0].overdraftLimit).toBe(0.0);
    });

    it("should call GET /accounts", async () => {
      const mock = new MockTransport(makeAccountListResponse());
      const resource = new AccountsResource(mock.transport);

      await resource.balances();

      expect(mock.callCount).toBe(1);
      expect(mock.lastCall.method).toBe("GET");
      expect(mock.lastCall.path).toBe("/accounts");
      expect(mock.lastCall.options).toBeUndefined();
    });

    it("should include recent activity in balances response", async () => {
      const fixture = makeAccountListResponse();
      const mock = new MockTransport(fixture);
      const resource = new AccountsResource(mock.transport);

      const result = await resource.balances();
      const activity = result.items[0].recentActivity;

      expect(activity).toHaveLength(1);
      expect(activity[0].id).toBe("act-001");
      expect(activity[0].amount).toBe(-48.5);
      expect(activity[0].operation).toBe("purchase");
    });
  });

  describe("transactions()", () => {
    it("should return transactions on success", async () => {
      const fixture = makeAccountTransactionPage();
      const mock = new MockTransport(fixture);
      const resource = new AccountsResource(mock.transport);

      const result: AccountTransactionPage = await resource.transactions();

      expect(result.totalCount).toBe(1);
      expect(result.items).toHaveLength(1);
      expect(result.items[0].id).toBe("tx-001");
      expect(result.items[0].category).toBe("shop");
      expect(result.items[0].amount).toBe(-48.5);
      expect(result.items[0].orderId).toBe("ord-456");
    });

    it("should call GET /accounts/transactions with default pagination", async () => {
      const mock = new MockTransport(makeAccountTransactionPage());
      const resource = new AccountsResource(mock.transport);

      await resource.transactions();

      expect(mock.callCount).toBe(1);
      expect(mock.lastCall.method).toBe("GET");
      expect(mock.lastCall.path).toBe("/accounts/transactions");
      expect(mock.lastCall.options?.params).toEqual({
        limit: 50,
        offset: 0,
      });
    });

    it("should pass all filter options", async () => {
      const mock = new MockTransport(makeAccountTransactionPage());
      const resource = new AccountsResource(mock.transport);

      await resource.transactions({
        currency: "USD",
        category: ["shop", "funding"],
        search: "Steam",
        transactionId: "txn-abc-123",
        orderId: "ord-456",
        dateFrom: "2026-01-01",
        dateTo: "2026-03-01",
        limit: 25,
        offset: 10,
      });

      expect(mock.lastCall.options?.params).toEqual({
        currency: "USD",
        category: ["shop", "funding"],
        search: "Steam",
        transactionId: "txn-abc-123",
        orderId: "ord-456",
        dateFrom: "2026-01-01",
        dateTo: "2026-03-01",
        limit: 25,
        offset: 10,
      });
    });

    it("should handle no options (undefined)", async () => {
      const mock = new MockTransport(makeAccountTransactionPage());
      const resource = new AccountsResource(mock.transport);

      await resource.transactions(undefined);

      expect(mock.lastCall.options?.params).toEqual({
        limit: 50,
        offset: 0,
      });
    });
  });
});

File: sdk/approute-sdk/golang/accounts_test.go
package approute_test

import (
	"errors"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"

	approute "github.com/approute/public-api-sdk-go"
	"github.com/approute/public-api-sdk-go/apierror"
	"github.com/approute/public-api-sdk-go/enum"
	"github.com/approute/public-api-sdk-go/model"
	"github.com/approute/public-api-sdk-go/resource"
)

func TestAccountsResource_Balances(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		assertEqual(t, "Method", r.Method, "GET")
		assertEqual(t, "Path", r.URL.Path, "/accounts")
		w.Header().Set("Content-Type", "application/json")
		w.Write(successEnvelope(t, newAccountListResponse(
			newAccount(),
			newAccount(func(a *model.Account) {
				a.Currency = "EUR"
				a.Balance = 75.25
				a.Available = 75.25
				a.OverdraftLimit = 0
				a.RecentActivity = []model.AccountActivity{}
			}),
		)))
	}))
	defer srv.Close()

	client := approute.NewClient("test-key", approute.WithBaseURL(srv.URL), approute.WithMaxRetries(0))
	result, err := client.Accounts.Balances(t.Context())
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	if len(result.Items) != 2 {
		t.Fatalf("expected 2 accounts, got %d", len(result.Items))
	}
	assertEqual(t, "Items[0].Currency", result.Items[0].Currency, "USD")
	assertEqual(t, "Items[0].Balance", result.Items[0].Balance, 150.50)
	assertEqual(t, "Items[0].Available", result.Items[0].Available, 120.00)
	assertEqual(t, "Items[1].Currency", result.Items[1].Currency, "EUR")

	// Verify recent activity
	if len(result.Items[0].RecentActivity) != 1 {
		t.Fatalf("expected 1 activity for USD account, got %d", len(result.Items[0].RecentActivity))
	}
	assertEqual(t, "RecentActivity[0].ID", result.Items[0].RecentActivity[0].ID, "act-1")
	assertEqual(t, "RecentActivity[0].Amount", result.Items[0].RecentActivity[0].Amount, -10.00)
}

func TestAccountsResource_Transactions(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		assertEqual(t, "Method", r.Method, "GET")
		assertEqual(t, "Path", r.URL.Path, "/accounts/transactions")

		q := r.URL.Query()
		if q.Get("limit") == "" {
			t.Error("expected limit parameter")
		}
		if q.Get("offset") == "" {
			t.Error("expected offset parameter")
		}

		w.Header().Set("Content-Type", "application/json")
		w.Write(successEnvelope(t, newAccountTransactionPage()))
	}))
	defer srv.Close()

	client := approute.NewClient("test-key", approute.WithBaseURL(srv.URL), approute.WithMaxRetries(0))
	result, err := client.Accounts.Transactions(t.Context(), nil)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	assertEqual(t, "TotalCount", result.TotalCount, 1)
	if len(result.Items) != 1 {
		t.Fatalf("expected 1 transaction, got %d", len(result.Items))
	}
	assertEqual(t, "Items[0].ID", result.Items[0].ID, "txn-1")
	assertEqual(t, "Items[0].Category", string(result.Items[0].Category), "shop")
	assertEqual(t, "Items[0].Amount", result.Items[0].Amount, -10.00)
}

func TestAccountsResource_Transactions_WithOptions(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		q := r.URL.Query()
		assertEqual(t, "currency", q.Get("currency"), "USD")
		assertEqual(t, "category", q.Get("category"), "shop,funding")
		assertEqual(t, "search", q.Get("search"), "test-search")
		assertEqual(t, "limit", q.Get("limit"), "25")
		assertEqual(t, "offset", q.Get("offset"), "10")

		if q.Get("dateFrom") == "" {
			t.Error("expected dateFrom parameter")
		}
		if q.Get("dateTo") == "" {
			t.Error("expected dateTo parameter")
		}

		w.Header().Set("Content-Type", "application/json")
		w.Write(successEnvelope(t, newAccountTransactionPage()))
	}))
	defer srv.Close()

	from := time.Date(2026, 3, 1, 0, 0, 0, 0, time.UTC)
	to := time.Date(2026, 3, 5, 0, 0, 0, 0, time.UTC)

	client := approute.NewClient("test-key", approute.WithBaseURL(srv.URL), approute.WithMaxRetries(0))
	_, err := client.Accounts.Transactions(t.Context(), &resource.TransactionListOptions{
		Currency: "USD",
		Category: []enum.BalanceCategory{enum.BalanceShop, enum.BalanceFunding},
		Search:   "test-search",
		Limit:    25,
		Offset:   10,
		DateFrom: &from,
		DateTo:   &to,
	})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestAccountsResource_Balances_Unauthorized(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(401)
		w.Header().Set("Content-Type", "application/json")
		w.Write(errorEnvelope(t, "UNAUTHORIZED", "Invalid API key", "t-acc-err-1"))
	}))
	defer srv.Close()

	client := approute.NewClient("bad-key", approute.WithBaseURL(srv.URL), approute.WithMaxRetries(0))
	_, err := client.Accounts.Balances(t.Context())
	if err == nil {
		t.Fatal("expected error, got nil")
	}

	var unauth *apierror.UnauthorizedError
	if !errors.As(err, &unauth) {
		t.Fatalf("expected *UnauthorizedError, got %T: %v", err, err)
	}
	assertEqual(t, "Code", unauth.Code, "UNAUTHORIZED")
	assertEqual(t, "TraceID", unauth.TraceID, "t-acc-err-1")
}
